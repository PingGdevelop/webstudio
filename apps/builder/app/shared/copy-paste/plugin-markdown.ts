import store from "immerhin";
import { gfm } from "micromark-extension-gfm";
import { fromMarkdown } from "mdast-util-from-markdown";
import type { Instance, Prop } from "@webstudio-is/project-build";
import type { Root } from "mdast-util-from-markdown/lib";
import { utils } from "@webstudio-is/project";
import { nanoid } from "nanoid";
import {
  createInstancesIndex,
  findClosestDroppableTarget,
  insertInstanceMutable,
} from "../tree-utils";
import {
  instancesIndexStore,
  propsStore,
  rootInstanceContainer,
  selectedInstanceIdStore,
} from "../nano-states";

const micromarkOptions = { extensions: [gfm()] };

export const mimeType = "text/plain";

// @todo List, ListItem, Definition, Code, Blockquote, Break, Image
const astTypeComponentMap: Record<string, Instance["component"]> = {
  paragraph: "Paragraph",
  heading: "Heading",
  strong: "Bold",
  emphasis: "Italic",
  link: "RichTextLink",
  image: "Image",
};

type Options = { generateId?: typeof nanoid };

const toInstancesData = (
  ast: { children: Root["children"] },
  options: Options = {},
  props: Array<Prop> = []
): { instances: Instance["children"]; props: Array<Prop> } => {
  const { generateId = nanoid } = options;
  const instances: Instance["children"] = [];

  for (const child of ast.children) {
    const component = astTypeComponentMap[child.type];
    if (component) {
      const instance = utils.tree.createInstance({
        id: generateId(),
        component,
        children:
          "children" in child
            ? toInstancesData(child, options, props).instances
            : [],
      });
      instances.push(instance);
      if (child.type === "heading") {
        props.push({
          id: generateId(),
          type: "string",
          name: "tag",
          instanceId: instance.id,
          value: `h${child.depth}`,
        });
      }
      if (child.type === "link") {
        props.push({
          id: generateId(),
          type: "string",
          name: "href",
          instanceId: instance.id,
          value: child.url,
        });
      }
      if (child.type === "image") {
        props.push({
          id: generateId(),
          type: "string",
          name: "src",
          instanceId: instance.id,
          value: child.url,
        });
      }
      if ("title" in child && child.title) {
        props.push({
          id: generateId(),
          type: "string",
          name: "title",
          instanceId: instance.id,
          value: child.title,
        });
      }

      if ("alt" in child && child.alt) {
        props.push({
          id: generateId(),
          type: "string",
          name: "alt",
          instanceId: instance.id,
          value: child.alt,
        });
      }
    }

    if (child.type === "text") {
      instances.push({
        type: "text",
        value: child.value,
      });
    }
  }
  return { instances, props };
};

export const parse = (clipboardData: string, options?: Options) => {
  const ast = fromMarkdown(clipboardData, micromarkOptions);
  if (ast.children.length === 0) {
    return;
  }
  console.log(JSON.stringify(ast, null, 2));
  return toInstancesData(ast, options);
};

export const onPaste = (clipboardData: string) => {
  const data = parse(clipboardData);
  if (data === undefined) {
    return;
  }

  const dropTarget = findClosestDroppableTarget(
    instancesIndexStore.get(),
    selectedInstanceIdStore.get()
  );
  store.createTransaction(
    [rootInstanceContainer, propsStore],
    (rootInstance, props) => {
      const instancesIndex = createInstancesIndex(rootInstance);
      for (const instance of data.instances) {
        insertInstanceMutable(instancesIndex, instance, dropTarget);
      }
      for (const prop of data.props) {
        props.set(prop.id, prop);
      }
    }
  );
};
