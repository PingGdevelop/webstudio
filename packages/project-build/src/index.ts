export * from "./types";
export * from "./shared/pages-utils";
export * from "./shared/folder-utils";
