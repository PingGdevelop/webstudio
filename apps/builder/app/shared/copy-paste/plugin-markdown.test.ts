import { describe, expect } from "@jest/globals";
import { Instance, Prop } from "@webstudio-is/project-build";
import { parse } from "./plugin-markdown";

const parseInstanceData = (data?: {
  instances: Instance["children"];
  props: Array<Prop>;
}) => {
  if (data === undefined) {
    return;
  }
  const { instances, props } = data;
  for (const instance of instances) {
    if ("id" in instance) {
      Instance.parse(instance);
    }
    if ("children" in instance) {
      parseInstanceData({ instances: instance.children, props });
    }
  }
  for (const prop of props) {
    Prop.parse(prop);
  }
  return { instances, props };
};

const options = { generateId: () => "123" };

describe("Plugin Markdown", () => {
  test("paragraph", () => {
    expect(parseInstanceData(parse("xyz", options))).toMatchSnapshot();
  });

  test("h1", () => {
    expect(parseInstanceData(parse("# heading", options))).toMatchSnapshot();
  });

  test("h6", () => {
    expect(
      parseInstanceData(parse("###### heading", options))
    ).toMatchSnapshot();
  });

  test("bold 1", () => {
    expect(parseInstanceData(parse("__bold__", options))).toMatchSnapshot();
  });

  test("bold 2", () => {
    expect(parseInstanceData(parse("**bold**", options))).toMatchSnapshot();
  });

  test("italic 1", () => {
    expect(parseInstanceData(parse("_italic_", options))).toMatchSnapshot();
  });

  test("italic 2", () => {
    expect(parseInstanceData(parse("*italic*", options))).toMatchSnapshot();
  });

  test("link", () => {
    expect(
      parseInstanceData(parse('[link](/uri "Title")', options))
    ).toMatchSnapshot();
  });

  test("image", () => {
    expect(
      parseInstanceData(parse('![foo](/url "title")', options))
    ).toMatchSnapshot();
  });
});
