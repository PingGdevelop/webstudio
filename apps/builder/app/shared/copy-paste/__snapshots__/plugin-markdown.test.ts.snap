// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Plugin Markdown blockquote 1`] = `
{
  "instances": [
    {
      "children": [
        {
          "children": [
            {
              "type": "text",
              "value": "bar",
            },
          ],
          "component": "Paragraph",
          "id": "123",
          "type": "instance",
        },
      ],
      "component": "Blockquote",
      "id": "123",
      "type": "instance",
    },
  ],
  "props": [],
}
`;

exports[`Plugin Markdown bold 1 1`] = `
{
  "instances": [
    {
      "children": [
        {
          "children": [
            {
              "type": "text",
              "value": "bold",
            },
          ],
          "component": "Bold",
          "id": "123",
          "type": "instance",
        },
      ],
      "component": "Paragraph",
      "id": "123",
      "type": "instance",
    },
  ],
  "props": [],
}
`;

exports[`Plugin Markdown bold 2 1`] = `
{
  "instances": [
    {
      "children": [
        {
          "children": [
            {
              "type": "text",
              "value": "bold",
            },
          ],
          "component": "Bold",
          "id": "123",
          "type": "instance",
        },
      ],
      "component": "Paragraph",
      "id": "123",
      "type": "instance",
    },
  ],
  "props": [],
}
`;

exports[`Plugin Markdown h1 1`] = `
{
  "instances": [
    {
      "children": [
        {
          "type": "text",
          "value": "heading",
        },
      ],
      "component": "Heading",
      "id": "123",
      "type": "instance",
    },
  ],
  "props": [
    {
      "id": "123",
      "instanceId": "123",
      "name": "tag",
      "type": "string",
      "value": "h1",
    },
  ],
}
`;

exports[`Plugin Markdown h6 1`] = `
{
  "instances": [
    {
      "children": [
        {
          "type": "text",
          "value": "heading",
        },
      ],
      "component": "Heading",
      "id": "123",
      "type": "instance",
    },
  ],
  "props": [
    {
      "id": "123",
      "instanceId": "123",
      "name": "tag",
      "type": "string",
      "value": "h6",
    },
  ],
}
`;

exports[`Plugin Markdown hard line break 1`] = `
{
  "instances": [
    {
      "children": [
        {
          "type": "text",
          "value": "foo",
        },
        {
          "type": "text",
          "value": "baz",
        },
      ],
      "component": "Paragraph",
      "id": "123",
      "type": "instance",
    },
  ],
  "props": [],
}
`;

exports[`Plugin Markdown image 1`] = `
{
  "instances": [
    {
      "children": [
        {
          "children": [],
          "component": "Image",
          "id": "123",
          "type": "instance",
        },
      ],
      "component": "Paragraph",
      "id": "123",
      "type": "instance",
    },
  ],
  "props": [
    {
      "id": "123",
      "instanceId": "123",
      "name": "src",
      "type": "string",
      "value": "/url",
    },
    {
      "id": "123",
      "instanceId": "123",
      "name": "title",
      "type": "string",
      "value": "title",
    },
    {
      "id": "123",
      "instanceId": "123",
      "name": "alt",
      "type": "string",
      "value": "foo",
    },
  ],
}
`;

exports[`Plugin Markdown italic 1 1`] = `
{
  "instances": [
    {
      "children": [
        {
          "children": [
            {
              "type": "text",
              "value": "italic",
            },
          ],
          "component": "Italic",
          "id": "123",
          "type": "instance",
        },
      ],
      "component": "Paragraph",
      "id": "123",
      "type": "instance",
    },
  ],
  "props": [],
}
`;

exports[`Plugin Markdown italic 2 1`] = `
{
  "instances": [
    {
      "children": [
        {
          "children": [
            {
              "type": "text",
              "value": "italic",
            },
          ],
          "component": "Italic",
          "id": "123",
          "type": "instance",
        },
      ],
      "component": "Paragraph",
      "id": "123",
      "type": "instance",
    },
  ],
  "props": [],
}
`;

exports[`Plugin Markdown link 1`] = `
{
  "instances": [
    {
      "children": [
        {
          "children": [
            {
              "type": "text",
              "value": "link",
            },
          ],
          "component": "RichTextLink",
          "id": "123",
          "type": "instance",
        },
      ],
      "component": "Paragraph",
      "id": "123",
      "type": "instance",
    },
  ],
  "props": [
    {
      "id": "123",
      "instanceId": "123",
      "name": "href",
      "type": "string",
      "value": "/uri",
    },
    {
      "id": "123",
      "instanceId": "123",
      "name": "title",
      "type": "string",
      "value": "Title",
    },
  ],
}
`;

exports[`Plugin Markdown paragraph 1`] = `
{
  "instances": [
    {
      "children": [
        {
          "type": "text",
          "value": "xyz",
        },
      ],
      "component": "Paragraph",
      "id": "123",
      "type": "instance",
    },
  ],
  "props": [],
}
`;

exports[`Plugin Markdown soft line break 1`] = `
{
  "instances": [
    {
      "children": [
        {
          "type": "text",
          "value": "foo
baz",
        },
      ],
      "component": "Paragraph",
      "id": "123",
      "type": "instance",
    },
  ],
  "props": [],
}
`;
